### DISCUSSION TITLE
How to type yield fixtures

### DISCUSSION THREAD

**Discussion by OlgaPaw** [pytest-dev/pytest#7809] [General] [answered]
Hello!

I have an example pytest fixture that uses `yield`

```
@pytest.fixture()
def foo() -> str:
    yield "foo"
```

Mypy gives an error `The return type of a generator function should be "Generator" or one of its supertypes`<

How can I type that fixture, so mypy understands it?

**Comment by The-Compiler** [comment]
I'd expect `Iterator[str]`, see https://mypy.readthedocs.io/en/stable/kinds_of_types.html#generators

**Comment by bluetech** [comment, marked as answer]
`Iterator[str]` is actually not strong enough; the correct incantation is

```py
from typing import Generator

@pytest.fixture()
def foo() -> Generator[str, None, None]:
    yield "foo"
```

I'm hoping one day generics could have defaults, in which case the type would be just `Generator[str]`, but for now the full thing is needed.

**Reply by graingert** [reply]
fyi you can parametrize type aliases:

```python
from typing import Generator, TypeVar
import pytest

T = TypeVar("T")

YieldFixture = Generator[T, None, None]


@pytest.fixture
def foo() -> YieldFixture[str]:
    yield "foo"
```

Live demo: https://mypy-play.net/?mypy=latest&python=3.8&flags=strict
